File: 647.palindromic-substrings.js
/*
 * @lc app=leetcode id=647 lang=javascript
 *
 * [647] Palindromic Substrings
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function(s) {
  let total = s.length

  // create hash map with char: indexes 'a': [0,2,6]
  let map = {}
  let v
  for (let i=0; i<s.length; i++) {
    v = s[i]
    map[v] = [...(map[v] || []), i]
  }

  // iterate over each index, and subsequent indexes,
  Object.entries(map).map(([char,indexes])=>{
    // and see if the substring between them is a palindrome
    for (let i=0; i < indexes.length; i++) {
      for (let j=i+1; j < indexes.length; j++) {
        let substr = s.substring(indexes[i], indexes[j]+1)
        if (substr === [...substr].reverse().join(''))
          total++
      }
    }
  })

  return total
};
// @lc code=end

File: 647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    def countSubstrings(self, s: str) -> int:
        count = 0
        for i in range(0, len(s))[::-1]:
            c1 = s[i]
            lastIndexes = [j+i for j, c2 in enumerate(s[i:]) if c1 == c2]
            for j in lastIndexes:
                substring = s[i:j+1]
                if self.isPalindrome(substring):
                    count += 1
        return count

    def isPalindrome(self, s: str) -> bool:
        return s == s[::-1]

# @lc code=end

